clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -459,7 +467,506 @@ void SemaHLSL::handleResourceClassAttr(Decl *D, const ParsedAttr &AL) {
+
+  // Samplers, UAVs, and SRVs are VarDecl types
+  VarDecl *TheVarDecl = dyn_cast<VarDecl>(TheDecl);
+  // Cbuffers and Tbuffers are HLSLBufferDecl types
+  HLSLBufferDecl *CBufferOrTBuffer = dyn_cast<HLSLBufferDecl>(TheDecl);
+
+  // exactly one of these two types should be set
+  assert(((TheVarDecl && !CBufferOrTBuffer) ||
+          (!TheVarDecl && CBufferOrTBuffer)) &&
+         "either TheVarDecl or CBufferOrTBuffer should be set");
+
+  RegisterBindingFlags Flags = HLSLFillRegisterBindingFlags(S, TheDecl);
+  assert((int)Flags.Other + (int)Flags.Resource + (int)Flags.Basic +
+                 (int)Flags.UDT ==
+             1 &&
+         "only one resource analysis result should be expected");
+
+  int regType = getRegisterTypeIndex(Slot);
+
+  // first, if "other" is set, emit an error
+  if (Flags.Other) {
+    if (regType == RegisterType::I) {
+      S.Diag(TheDecl->getLocation(),
+             diag::warn_hlsl_deprecated_register_type_i);
+      return;
+    }
+    S.Diag(ArgLoc, diag::err_hlsl_binding_type_mismatch) << regType;
+    return;
+  }
+
+  // next, if multiple register annotations exist, check that none conflict.
+  ValidateMultipleRegisterAnnotations(S, TheDecl, Slot);
+
+  // next, if resource is set, make sure the register type in the register
+  // annotation is compatible with the variable's resource type.
+  if (Flags.Resource) {
+    if (regType == RegisterType::I) {
+      S.Diag(TheDecl->getLocation(),
+             diag::warn_hlsl_deprecated_register_type_i);
+      return;
+    }
+    const HLSLResourceAttr *resAttr =
+        getHLSLResourceAttrFromEitherDecl(TheVarDecl, CBufferOrTBuffer);
+    const HLSLResourceClassAttr *resClassAttr =
+        getHLSLResourceClassAttrFromEitherDecl(TheVarDecl, CBufferOrTBuffer);
+    assert(resAttr && resClassAttr &&
+           "any decl that set the resource flag on analysis should "
+           "have a resource attribute and resource class attribute attached.");
+    const llvm::hlsl::ResourceClass DeclResourceClass =
+        resClassAttr->getResourceClass();
+
+    switch (DeclResourceClass) {
+    case llvm::hlsl::ResourceClass::SRV:
+      if (regType != RegisterType::SRV)
+        S.Diag(TheDecl->getLocation(), diag::err_hlsl_binding_type_mismatch)
+            << regType;
+      break;
+    case llvm::hlsl::ResourceClass::UAV:
+      if (regType != RegisterType::UAV)
+        S.Diag(TheDecl->getLocation(), diag::err_hlsl_binding_type_mismatch)
+            << regType;
+      break;
+    case llvm::hlsl::ResourceClass::CBuffer:
+      if (regType != RegisterType::CBuffer)
+        S.Diag(TheDecl->getLocation(), diag::err_hlsl_binding_type_mismatch)
+            << regType;
+      break;
+    case llvm::hlsl::ResourceClass::Sampler:
+      if (regType != RegisterType::Sampler)
+        S.Diag(TheDecl->getLocation(), diag::err_hlsl_binding_type_mismatch)
+            << regType;
+      break;
+    }
```

Comment:
Something like this might better show the intent that this is looking for a mismatch between the register type implied by the resource class and the specified register type.

```suggestion
    static RegisterType ExpectedRegisterTypesForResourceClass[] = {
        RegisterType::SRV,
        RegisterType::UAV,
        RegisterType::CBuffer,
        RegisterType::Sampler,
        };
    assert(DeclResourceClass < _countof(ExpectedRegisterTypesForResourceClass));
    
    RegisterType ExpectedRegisterType = ExpectedRegisterTypesForResourceClass[DeclResourceClass];
    if (regType != ExpectedRegisterType) {
        S.Diag(TheDecl->getLocation(), diag::err_hlsl_binding_type_mismatch)
            << regType;
    }
```

Or you could have a `GetRegisterTypeForResourceClass(...)` function that has a switch in it rather than the lookup table if you prefer.  But IMO it'd be good to try and get something in the code that looks like a table of expected results, rather than combining the information about expected register types with checking that they match and emitting the diagnostic about it.